Adding support for subreddit reports and mod queues

devmedoo opened the pull request:
Added reports and mod queue fetching with streaming support.
Took the liberty to remove ReportReasons (reddit API's `report_reasons`) as it always returns "This attribute is deprecated. Please use mod_reports and user_reports instead." now.

devmedoo committed:
Added subreddits' reports & mod queue w/streaming.
Removed report_reasons due to API duprecation.

thecsw commented on models/modqueue.go:
```diff
@@ -0,0 +1,11 @@
+func (mql ModQueueListingChild) GetId() string {
+	if mql.GetKind() == "t1" {
+		return mql.Data.(CommentListingDataChildrenData).Name
+	} else {
```
no need to have an else here, the first return will already exit the scope

devmedoo replied:
I completely agree, thanks.

thecsw commented on models/post_struct.go:
```diff
@@ -98,7 +98,6 @@ type PostListingChildData struct {
 	LinkFlairBackgroundColor   string             `json:"link_flair_background_color"`
 	Id                         string             `json:"id"`
 	IsRobotIndexable           bool               `json:"is_robot_indexable"`
-	ReportReasons              string             `json:"report_reasons"`
```
Thanks for the comment regarding this deprecated feature! I agree that it should not be used, however, I am not sure how many programs out there might inadvertently use it. In spirits of not breaking compatibility, let's just leave it there. (we can remove it and bump mira to a new major version)

devmedoo replied:
Added it back and attached a comment (maybe it can show up on some IDEs/godocs?)

thecsw commented on models/reports.go:
```diff
@@ -0,0 +1,11 @@
+func (rl ReportListingChild) GetId() string {
+	if rl.GetKind() == "t1" {
+		return rl.Data.(CommentListingDataChildrenData).Name
+	} else {
```
we can get rid of the else and have the second return in the scope of the function

thecsw commented on models/user_reports.go:
```diff
@@ -0,0 +1,5 @@
+package models
+
+type UserReport struct {
+	Reason string `json:"0"`
```
interesting, are user reports objects with `0` as their key or are they an array of one element?

devmedoo replied:
```
{
  "user_reports": [
    [
      "testing tatata (mods please ignore this)",
      1,
      false,
      true
    ]
  ]
}
```
This is an example report I made through reddit.

thecsw replied:
hmm, that looks like an array to me. So just as a recap. This is how we unmarshall `user_reports`
```go
UserReports                []UserReport       `json:"user_reports"`
```
So `UserReports` is a slice of `UserReport`, where  each `UserReport` is a struct
```go
type UserReport struct {
 	Reason string `json:"0"`
}
```
Then shouldn't each `Reason` be an array of 4 elements (maybe it's an object if we look at the Reddit API docs?) with the example you gave above?
```json
[
      "testing tatata (mods please ignore this)",
      1,
      false,
      true
]
```
This denotes an array or an object? with types of `string, int, bool, bool`

devmedoo replied:
It does denote an array. I am unable to find out what `int, bool, bool`  refer to. Is there any reddit API documentation other than [this](https://www.reddit.com/dev/api)?

thecsw replied:
@devmedoo Weird. The link you sent is the reddit doc that I based mira on. I believe it should be working then. Consider this resolved. Thanks

devmedoo replied:
I am unable to find any information regarding outputs of API calls on that page. It lists endpoints and their brief description and inputs. Am I missing something?

thecsw replied:
Have you tried manually making a call with `MiraRequest` and seeing what Reddit returns for that endpoint and element?

UPD. We could probe the endpoints and API until we find a structure that we're comfortable with.

devmedoo replied:
Did some digging. [Found](https://www.reddit.com/r/redditdev/comments/olqo5s/comment/h5g71gf/?utm_source=reddit&utm_medium=web2x&context=3) a reddit admin explaining each array element. 

I don't think my approach of using array index will work with the default json library as it is an array that needs to be unmarshalled into a struct. 
Maybe I should change the UserReports field type to `[]interface{}` then add a method to return user reports as an array of structs using casting?

thecsw replied:
Nice find! Why can't we declare it as an array of structs and unmarshall it there? Even if most of the time it will be an array of size 1. Sounds like a good idea

devmedoo replied:
The ` [ ] ` in [the reddit API response](https://github.com/thecsw/mira/pull/12#discussion_r801937398) denotes its type as array. As far as I know `json.Unmarshal` default behavior does not convert JSON arrays to structs. [This is a similar example with a much better explanation.](https://stackoverflow.com/a/42378906)

thecsw replied:
@devmedoo run this code, where we unmarshall a JSON array into structs

I made a playground example -> https://go.dev/play/p/4c2OSUAOTKA

```go
package main

import (
	"encoding/json"
	"fmt"
)

type MyStruct struct {
	Value1 string `json:"value_one"`
	Value2 int    `json:"value_two"`
	Value3 bool   `json:"value_three"`
}

func main() {
	vals := make([]MyStruct, 0, 10)

	json.Unmarshal([]byte(`[
{
  "value_one": "hello",
  "value_two": 37,
  "value_three": true
},
{
  "value_one": "this works?",
  "value_two": 9999,
  "value_three": false
}
]`), &vals)

	fmt.Println(vals)
}
```

As expected, it prints `[{hello 37 true} {this works? 9999 false}]`

devmedoo replied:
```
"user_reports": [
    [

    ]
]
```

It is an array of arrays. Your example is an array of objects. In JSON, square brackets `[]` denote arrays and curly brackets `{}` denote objects. Objects have a defined index (i.e: `"value_one"`). While arrays do not have a defined index and usually is iterated through numerically in programs. Converting your example JSON string to one that is an array of arrays instead of an array of objects in [this playground](https://go.dev/play/p/j5GPHhInzz-) does not provide the desired output. What I suggested in my previous comment works like [this](https://go.dev/play/p/X2hr3ZZRXRO).

thecsw replied:
That makes sense. Thanks for clarifying that up! I think your approach here is the simplest one, therefore most optimal, and nice to read. Approved

devmedoo replied:
On second thought, since I can't change the data type for what should be
```go
UserReports                []interface{}       `json:"user_reports" 
````
into an array of UserReport during runtime, maybe I should just expose the report reason, count, snooze status and can snooze through methods of ReportListingChild?

thecsw commented on streaming.go:
```diff
@@ -82,6 +82,88 @@ func (c *Reddit) StreamSubmissions() (<-chan models.PostListingChild, error) {
+			for i := range new {
+				ret <- new[len(new)-i-1]
+			}
+			time.Sleep(c.Stream.PostListInterval * time.Second)
```
We can take the liberty and maybe add a new interval member to the stream struct that would be used specifically for modqueue streaming

devmedoo replied:
Added new intervals for both modqueue and reports.

thecsw replied:
Great! When you push the changes, I'll look over and it should get merged. Thanks

thecsw commented on models/post_struct.go:
```diff
@@ -47,7 +47,7 @@ type PostListingChildData struct {
 	AuthorFlairTemplateId      string             `json:"author_flair_template_id"`
 	IsOriginalContent          bool               `json:"is_original_content"`
-	UserReports                []string           `json:"user_reports"`
+	UserReports                []UserReport       `json:"user_reports"`
```
This is a little bit of a breaking change but with the update requested here, I think it's okay

thecsw commented on models/reports_struct.go:
```diff
@@ -0,0 +1,19 @@
+
+type ReportListingChild struct {
+	Kind   string      `json:"kind"`
+	Data   interface{} `json:"data"`
```
Do we just not care about this field? (totally ok if we don't)

devmedoo replied:
I apologise if it is obvious, but I am not sure which field you are referring to.
If it is Kind, then we do care about it, because it is used to identify whether the Report/Modqueue entry is a thread or a comment.
If it is Data, then we do care about it, I opted to choose `interface{}` as the data type as it will be later cast to a `CommentListingDataChildrenData` if it is a comment and `PostListingChildData` if it is a thread.

thecsw replied:
I see! When I add a comment to a line, I guess Github doesn't really highlight it. It's usually the very last line in the code review window, so I was referring to `Data`. Good stuff, I see where the casting is done, it's good!

thecsw reviewed (changes requested):
Hey! Thanks for the PR. Great stuff. I left some comments regarding breaking changes, small go style requests, and your consultation regarding some of the things. Let me know what you think

devmedoo committed:
Added intervals for modqueue and reports.
Reverted breaking change of "report_reasons" removal.

devmedoo committed:
Added go.mod
Added tests for modqueue and reports
Fixed issues with modqueue and reports implementation

devmedoo committed:
Added go.mod
Added tests for modqueue and reports
Fixed issues with modqueue and reports implementation

devmedoo commented:
Ok, I think I figured it out. I added and ran tests to make sure everything is sound.

thecsw commented on models/post_struct.go:
```diff
@@ -98,6 +98,7 @@ type PostListingChildData struct {
 	LinkFlairBackgroundColor   string             `json:"link_flair_background_color"`
 	Id                         string             `json:"id"`
 	IsRobotIndexable           bool               `json:"is_robot_indexable"`
+	ReportReasons              string             `json:"report_reasons"` // WILL BE REMOVED: as reddit's API always returns "This attribute is deprecated. Please use mod_reports and user_reports instead." now
```
love it!

thecsw commented on models/reports.go:
```diff
@@ -1,11 +1,19 @@
+		report.NumOfReports = rl.Data.UserReports[i].([]interface{})[1].(float64)
+		report.SnoozeStatus = rl.Data.UserReports[i].([]interface{})[2].(bool)
+		report.CanSnooze = rl.Data.UserReports[i].([]interface{})[3].(bool)
+		reports = append(reports, report)
```
I like this casting, great job! It also simplified some getters and got rid of the if statement there as we verify the type before then

thecsw commented on models/modqueue_test.go:
```diff
@@ -0,0 +1,20 @@
+	"testing"
+)
+
+func TestGetModQueueListingId(t *testing.T) {
```
Good tests

thecsw commented on models/user_reports.go:
```diff
@@ -1,5 +1,8 @@
 package models
 
 type UserReport struct {
-	Reason string `json:"0"`
```
Great! This reads better

thecsw reviewed (approved):
@devmedoo Amazing PR! Great great job on this one. Approved and merging
